src/store/slices/balanceHistorySlice/balanceHistorySlice.ts

```diff
@@ -0,0 +1,216 @@
+import {
+  createAsyncThunk,
+  createSlice,
+  isPending,
+  isFulfilled,
+  isRejected,
+} from '@reduxjs/toolkit';
+import { DocumentData, QuerySnapshot } from 'firebase/firestore';
+
+import {
+  getFirestoreSubCollectionById,
+  setFirestoreSubCollectionById,
+  updateFirestoreDataById,
+} from 'api';
+import { BalanceOperation, FieldsTransaction, NamesDBCollection } from 'constant';
+import { handleError } from 'helpers';
+import { ThunkApiConfig, setErrorMessage } from 'store';
+import { ISetBalanceHistoryThunkProps, Transaction } from 'types';
+
+export interface IBalanceHistory {
+  transaction: Transaction;
+  loader: boolean;
+  totalBalance: number;
+  historyBalance: Transaction[] | [];
+  lastVisibleData: Nullable<QuerySnapshot>;
+}
+
+const initialState: IBalanceHistory = {
+  transaction: {
+    typeOperation: null,
+    currentBalance: 0,
+    amount: 0,
+    performerName: '',
+    dateOperation: new Date(),
+  },
+  loader: false,
+  totalBalance: 0,
+  historyBalance: [],
+  lastVisibleData: null,
+};
+
+export const changeBalanceThunk = createAsyncThunk<
+  number,
+  ISetBalanceHistoryThunkProps,
+  ThunkApiConfig
+>('balanceHistorySlice/changeBalance', async (data, { dispatch, rejectWithValue }) => {
+  try {
+    const { currentBalance, amount, typeOperation } = data.values;
+    let totalBalance = currentBalance;
+
+    const decrement = (): number => {
+      if (currentBalance < amount) {
+        dispatch(setErrorMessage('errorInsufficientFunds'));
+
+        return currentBalance;
+      }
+      totalBalance -= amount;
+
+      return totalBalance;
+    };
+
+    const increment = (): number => {
+      totalBalance += amount;
+
+      return totalBalance;
+    };
+
+    switch (typeOperation) {
+      case BalanceOperation.Accrue:
+        increment();
+        break;
+      case BalanceOperation.WriteOff:
+        decrement();
+        break;
+      default:
+        dispatch(setErrorMessage('errorOperation'));
+        break;
+    }
+
+    const value = {
+      id: data.id,
+      values: {
+        balance: totalBalance,
+      },
+    };
+
+    await updateFirestoreDataById(value, NamesDBCollection.Users);
+
+    return totalBalance;
+  } catch (error) {
+    const message = handleError(error as Error);
+
+    dispatch(setErrorMessage(message));
+
+    return rejectWithValue(message);
+  }
+});
```

Comment:
try to simplify your logic,
try using [increment/decrement](https://firebase.google.com/docs/firestore/manage-data/add-data?authuser=0#web-modular-api_13)